modules/add-contextual-data/add-contextual-data-template-selector.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2016 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "add-contextual-data-template-selector.h"
+#include "template/templates.h"
+#include "syslog-ng.h"
+#include "messages.h"
+
+typedef struct _AddContextualDataTemplateSelector
+{
+  AddContextualDataSelector super;
+  gchar *selector_template_string;
+  LogTemplate *selector_template;
+} AddContextualDataTemplateSelector;
+
+static gboolean
+_compile_selector_template(AddContextualDataTemplateSelector *self)
+{
+  GError *error = NULL;
+  if (!self->selector_template_string)
+    {
+      msg_error("No selector set.");
+      return FALSE;
+    }
+
+  if (!log_template_compile(self->selector_template, self->selector_template_string, &error))
+    {
+      msg_error("Failed to compile template",
+                evt_tag_str("template", self->selector_template_string),
+                evt_tag_str("error", error->message));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+_replace_template(LogTemplate **old_template, LogTemplate *new_template)
+{
+  log_template_unref(*old_template);
+  *old_template = log_template_ref(new_template);
+}
+
+static gboolean
+_init(AddContextualDataSelector *s)
+{
+  AddContextualDataTemplateSelector *self = (AddContextualDataTemplateSelector *)s;
+  return _compile_selector_template(self);
+}
+
+static gchar *
+_resolve(AddContextualDataSelector *s, LogMessage *msg)
+{
+  GString *selector_str = g_string_new(NULL);
```

Comment:
GString should be freed with `g_string_free()` (it has additional fields, not just a gchar buffer).